packages/lit-ssr/src/lib/render-lit-html.ts

```diff
@@ -0,0 +1,551 @@
+import {LitElementRenderer} from './lit-element-renderer.js';
+import {reflectedAttributeName} from './reflected-attributes.js';
+
+declare module 'parse5' {
+  interface DefaultTreeElement {
+    isDefinedCustomElement?: boolean;
+  }
+}
+
+Directive.prototype.resolve = function (
+  this: Directive,
+  _part: Part,
+  values: unknown[]
+) {
+  return this.render(...values);
+};
+
+const templateCache = new Map<
+  TemplateStringsArray,
+  {
+    ops: Array<Op>;
+    attrNames: Array<string>;
+  }
+>();
+
+/**
+ * Operation to output static text
+ */
+type TextOp = {
+  type: 'text';
+  value: string;
+};
+
+/**
+ * Operation to output dynamic text from the associated template result value
+ */
+type NodePartOp = {
+  type: 'node-part';
+  index: number;
+  useCustomElementInstance?: boolean;
+};
+
+/**
+ * Operation to output an attribute with bindings. Includes all bindings for an
+ * attribute, like an attribute template part or AttributeComitter.
+ */
+type AttributePartOp = {
+  type: 'attribute-part';
+  index: number;
+  name: string;
+  strings: Array<string>;
+  tagName: string;
+  useCustomElementInstance?: boolean;
+};
+
+/**
+ * Operator to create a custom element instance.
+ */
+type CustomElementOpenOp = {
+  type: 'custom-element-open';
+  tagName: string;
+  ctor: {new (): HTMLElement};
+  staticAttributes: Map<string, string>;
+};
+
+/**
+ * Operation to render a custom element's attributes. This is separate from
+ * `custom-element-open` because attribute/property parts go in between and need
+ * to run and be set on the instance before we render the element's final
+ * attributes.
+ */
+type CustomElementAttributesOp = {
+  type: 'custom-element-attributes';
+};
+
+/**
+ * Operation to render a custom element's children, usually its shadow root.
+ */
+type CustomElementChildrenOp = {
+  type: 'custom-element-children';
+};
+
+/**
+ * Operation to close a custom element so that its no longer available for
+ * bindings.
+ */
+type CustomElementClosedOp = {
+  type: 'custom-element-close';
+};
+
+type Op =
+  | TextOp
+  | NodePartOp
+  | AttributePartOp
+  | CustomElementOpenOp
+  | CustomElementAttributesOp
+  | CustomElementChildrenOp
+  | CustomElementClosedOp;
+
+const getTemplate = (result: TemplateResult) => {
+  const template = templateCache.get(result.strings);
+  if (template !== undefined) {
+    return template;
+  }
+  const {html, attrNames} = getTemplateHtml(result.strings, result._$litType$);
+  const ast = parseFragment(html, {
+    sourceCodeLocationInfo: true,
+  }) as DefaultTreeDocumentFragment;
+
+  const ops: Array<Op> = [];
+
+  /* The last offset of html written to the stream */
+  let lastOffset: number | undefined = 0;
+
+  /**
+   * Sets `lastOffset` to `offset`, skipping a range of characters. This is
+   * useful for skipping <slot>s and distributed nodes in flattened mode, or
+   * skipping and re-writting lit-html marker nodes.
+   */
+  const skipTo = (offset: number) => {
```

Comment:
Think it's more readable if these aren't embedded functions.

Reply:
They are helpers working on local state that they close over (`offset` & `ops`). The state can be passed in and back out and re-assigned, but it'd make all the call sites a lot wordier.

Reply:
indeed